pkg/kinflate/util.go

```diff
@@ -17,122 +17,232 @@ limitations under the License.
 }
 
-// loadManifestPkg loads a manifest file and parse it in to the Package object.
-func loadManifestPkg(filename string) (*manifest.Manifest, error) {
-	bytes, err := ioutil.ReadFile(filename)
+// ManifestToMap builds a map given the info in manifest.
+func ManifestToMap(manifest *manifest.Manifest,
+	into map[kutil.GroupVersionKindName]*unstructured.Unstructured,
+) (map[kutil.GroupVersionKindName]*unstructured.Unstructured, error) {
+	baseResouceMap := map[kutil.GroupVersionKindName]*unstructured.Unstructured{}
+	if into != nil {
+		baseResouceMap = into
+	}
+	err := populateResourceMap(manifest.Resources, baseResouceMap)
 	if err != nil {
 		return nil, err
 	}
-	var pkg manifest.Manifest
-	// TODO: support json
-	err = yaml.Unmarshal(bytes, &pkg)
-	return &pkg, err
-}
 
-func populateResourceMap(files []string, m map[groupVersionKindName][]byte, errOut io.Writer) error {
-	decoder := unstructured.UnstructuredJSONScheme
+	overlayResouceMap := map[kutil.GroupVersionKindName]*unstructured.Unstructured{}
+	err = populateResourceMap(manifest.Patches, overlayResouceMap)
+	if err != nil {
+		return nil, err
+	}
 
-	for _, file := range files {
-		content, err := ioutil.ReadFile(file)
-		if err != nil {
-			return err
+	// Strategic merge the resources exist in both base and overlay.
+	for gvkn, base := range baseResouceMap {
+		// Merge overlay with base resource.
+		if overlay, found := overlayResouceMap[gvkn]; found {
+			versionedObj, err := scheme.Scheme.New(gvkn.GVK)
+			if err != nil {
+				switch {
+				case runtime.IsNotRegisteredError(err):
+					return nil, fmt.Errorf("CRD and TPR are not supported now: %v", err)
+				default:
+					return nil, err
+				}
+			}
+			merged, err := strategicpatch.StrategicMergeMapPatch(
+				base.UnstructuredContent(),
+				overlay.UnstructuredContent(),
+				versionedObj)
+			if err != nil {
+				return nil, err
+			}
+			baseResouceMap[gvkn].Object = merged
+			delete(overlayResouceMap, gvkn)
 		}
+	}
 
-		// try converting to json, if there is a error, probably because the content is already json.
-		jsoncontent, err := yaml.YAMLToJSON(content)
-		if err != nil {
-			fmt.Fprintf(errOut, "error when trying to convert yaml to json: %v\n", err)
-		} else {
-			content = jsoncontent
+	// If there are resources in overlay that are not defined in base, just add it to base.
+	if len(overlayResouceMap) > 0 {
+		for gvkn, jsonObj := range overlayResouceMap {
+			baseResouceMap[gvkn] = jsonObj
 		}
+	}
 
-		obj, gvk, err := decoder.Decode(content, nil, nil)
-		if err != nil {
-			return err
-		}
-		accessor, err := meta.Accessor(obj)
-		if err != nil {
-			return err
-		}
-		name := accessor.GetName()
-		gvkn := groupVersionKindName{gvk: *gvk, name: name}
+	err = populateConfigMapAndSecretMap(manifest, baseResouceMap)
+	if err != nil {
+		return nil, err
+	}
+
+	transformers, err := DefaultTransformers(manifest)
+	if err != nil {
+		return nil, err
+	}
+	for _, transformer := range transformers {
```

Comment:
s/transformer/t/  (it's a short loop)

BTW, love how this turned out.  Good effort.